File: scripts/generate-featured.mjs
import { createHash } from 'node:crypto';
import { mkdir, readFile, rm, writeFile } from 'node:fs/promises';

const GAME_ID = 8694;
// Only submissions in these FNF mod-folder trees are eligible. The legacy
// category roots (43772 and 3833) remain deliberately unsupported.
const CATEGORY_ROOTS = [
  29202, // Base Game / V-Slice
  28367, // Psych Engine
  34764, // Codename Engine
  3827, // Executables
  43798, // P-Slice
  44037, // ALE Psych
  43850, // FPS Plus
  43788, // Psych Online
  43774 // Legacy Base/Full Mods (direct links only)
];
// These entries are engine distributions rather than playable mods.
const EXCLUDED_MOD_IDS = new Set([309789]);
const TOP_SUBS_URL = `https://gamebanana.com/apiv12/Game/${GAME_ID}/TopSubs`;
const PROFILE_URL = 'https://gamebanana.com/apiv11/Mod';
const MOD_INDEX_URL = 'https://gamebanana.com/apiv11/Mod/Index';
const FEATURED_PER_PERIOD = 5;
const ENGINE_BY_CATEGORY = {
  29202: { id: 'vslice', name: 'Base Game', icon: 'vslice.png', categoryName: 'Base Game Mod Folders' },
  28367: { id: 'psych', name: 'Psych Engine', icon: 'psych.png', categoryName: 'Psych Engine Mod Folders' },
  34764: { id: 'codename', name: 'Codename Engine', icon: 'codename.png', categoryName: 'Codename Engine Mod Folders' },
  3827: { id: 'executable', name: 'Executable', icon: 'exe.png', categoryName: 'Executable Mod Folders' },
  43798: { id: 'pslice', name: 'P-Slice', icon: 'pslice.png', categoryName: 'P-Slice Mod Folders' },
  44037: { id: 'alepsych', name: 'ALE Psych', icon: 'alepsych.png', categoryName: 'ALE Psych Mod Folders' },
  43850: { id: 'fpsplus', name: 'FPS Plus', icon: 'fpsplus.png', categoryName: 'FPS Plus Mod Folders' },
  43788: { id: 'psychonline', name: 'Psych Online', icon: 'psychonline.png', categoryName: 'Psych Online Mod Folders' },
  43774: { id: 'vslice', name: 'Base Game', icon: 'vslice.png', categoryName: 'Originals / Full Mods (Base)' }
};
const VSLICE_CATEGORY_ROOTS = CATEGORY_ROOTS.filter((categoryId) => ENGINE_BY_CATEGORY[categoryId]?.id === 'vslice');
const PERIODS = [
  ['today', 'day', 'Best of Today', 24 * 60 * 60],
  ['week', 'week', 'Best of This Week', 7 * 24 * 60 * 60],
  ['month', 'month', 'Best of This Month', 30 * 24 * 60 * 60],
  ['3month', 'three-months', 'Best of 3 Months', 90 * 24 * 60 * 60],
  ['6month', 'six-months', 'Best of 6 Months', 180 * 24 * 60 * 60],
  ['year', 'year', 'Best of This Year', 365 * 24 * 60 * 60],
  ['alltime', 'all-time', 'Best of All Time', 0]
];

function recordsFrom(response) {
  return Array.isArray(response?._aRecords) ? response._aRecords : [];
}

function imageUrl(mod) {
  if (mod._sImageUrl) return mod._sImageUrl;
  const image = mod._aPreviewMedia?._aImages?.[0];
  return image
    ? `${image._sBaseUrl}/${image._sFile}`
    : 'https://images.gamebanana.com/img/ss/mods/default.jpg';
}

async function fetchTopSubs() {
  const response = await fetch(TOP_SUBS_URL);
  if (!response.ok) throw new Error(`GameBanana returned ${response.status} for TopSubs`);

  const mods = await response.json();
  if (!Array.isArray(mods)) throw new Error('GameBanana returned an invalid TopSubs response');
  return mods;
}

async function fetchProfile(modId) {
  const response = await fetch(`${PROFILE_URL}/${modId}/ProfilePage`);
  if (!response.ok) throw new Error(`GameBanana returned ${response.status} for mod ${modId}`);
  return response.json();
}

async function fetchCategory(categoryId, sort, pageLimit) {
  const mods = [];
  for (let page = 1; page <= pageLimit; page++) {
    const params = new URLSearchParams({ _sSort: sort, _nPage: String(page), _nPerpage: '50' });
    params.set('_aFilters[Generic_Game]', String(GAME_ID));
    params.set('_aFilters[Generic_Category]', String(categoryId));
    const response = await fetch(`${MOD_INDEX_URL}?${params}`);
    if (!response.ok) throw new Error(`GameBanana returned ${response.status} for category ${categoryId}`);
    const records = recordsFrom(await response.json());
    mods.push(...records);
    if (records.length < 50) break;
  }
  return mods.map((mod) => ({ mod, categoryId }));
}

function score(mod) {
  return (mod._nLikeCount || 0) * 1_000_000 + (mod._nDownloadCount || 0) * 1_000 + (mod._nViewCount || 0);
}

function activeSince(mod, cutoff) {
  return (mod._tsDateAdded || 0) >= cutoff || (mod._tsDateUpdated || mod._tsDateModified || 0) >= cutoff;
}

function uniqueMods(mods) {
  return [...new Map(mods.map((entry) => [entry.mod._idRow, entry])).values()]
    .filter(({ mod }) => !EXCLUDED_MOD_IDS.has(mod._idRow));
}

function toFeaturedMod({ mod, profile, categoryId }) {
  const engine = ENGINE_BY_CATEGORY[categoryId];
  return {
    id: mod._idRow,
    title: mod._sName,
    author: mod._aSubmitter?._sName || 'Unknown',
    image: imageUrl(mod),
    likes: mod._nLikeCount || profile?._nLikeCount || 0,
    downloads: profile._nDownloadCount || 0,
    views: profile._nViewCount || 0,
    publishedAt: profile._tsDateAdded || 0,
    updatedAt: profile._tsDateUpdated || profile._tsDateModified || 0,
    url: mod._sProfileUrl || profile._sProfileUrl || `https://gamebanana.com/mods/${mod._idRow}`,
    engine: { id: engine.id, name: engine.name, icon: engine.icon },
    category: { id: categoryId, name: engine.categoryName }
  };
}

async function fetchFeaturedSources() {
  const [topSubs, recentGroups, allTimeGroups] = await Promise.all([
    fetchTopSubs(),
    Promise.all(CATEGORY_ROOTS.map((categoryId) => fetchCategory(categoryId, 'Generic_NewAndUpdated', 4))),
    Promise.all(CATEGORY_ROOTS.map((categoryId) => fetchCategory(categoryId, 'Generic_MostLiked', 1)))
  ]);
  const uniqueTopSubs = [...new Map(topSubs.map((mod) => [mod._idRow, mod])).values()]
    .filter((mod) => !EXCLUDED_MOD_IDS.has(mod._idRow));
  const profiles = new Map(
    await Promise.all(uniqueTopSubs.map(async (mod) => [mod._idRow, await fetchProfile(mod._idRow)]))
  );

  return { topSubs, profiles, recentGroups, allTimeGroups };
}

async function buildFeaturedData(categoryRoots, sources) {
  const nowSeconds = Math.floor(Date.now() / 1000);
  const { topSubs, profiles, recentGroups, allTimeGroups } = sources;
  const allowedCategoryIds = new Set(categoryRoots);

  const eligibleMods = topSubs.filter((topSub) => {
    if (EXCLUDED_MOD_IDS.has(topSub._idRow)) return false;
    const categoryId = profiles.get(topSub._idRow)?._aSuperCategory?._idRow;
    return allowedCategoryIds.has(categoryId);
  });
  const recentMods = uniqueMods(recentGroups.flat().filter(({ categoryId }) => allowedCategoryIds.has(categoryId)));
  const allTimeMods = uniqueMods(allTimeGroups.flat().filter(({ categoryId }) => allowedCategoryIds.has(categoryId)));
  // A featured mod belongs to its most immediate qualifying period only.
  // Keeping this set outside the period loop prevents duplicate cards across
  // Today, Week, Month, and the longer rankings.
  const featuredModIds = new Set();
  const rankings = PERIODS.map(([apiPeriod, id, label, seconds]) => {
    // TopSubs is the canonical ranking. It exposes only three entries per
    // period, so use the same unbalanced popularity ordering only to fill the
    // remaining two slots from whitelisted submissions.
    const primary = eligibleMods
      .filter((topSub) => topSub._sPeriod === apiPeriod)
      .filter((topSub) => !featuredModIds.has(topSub._idRow))
      .map((topSub) => ({ mod: topSub, profile: profiles.get(topSub._idRow), categoryId: profiles.get(topSub._idRow)._aSuperCategory._idRow }));
    const selectedIds = new Set([...featuredModIds, ...primary.map(({ mod }) => mod._idRow)]);
    const cutoff = nowSeconds - seconds;
    const fallback = (seconds ? recentMods.filter(({ mod }) => activeSince(mod, cutoff)) : allTimeMods)
      .filter(({ mod }) => !selectedIds.has(mod._idRow))
      .sort((left, right) => score(right.mod) - score(left.mod) || right.mod._idRow - left.mod._idRow)
      .slice(0, FEATURED_PER_PERIOD - primary.length)
      .map(({ mod, categoryId }) => ({ mod, profile: mod, categoryId }));
    const mods = [...primary, ...fallback];
    mods.forEach(({ mod }) => featuredModIds.add(mod._idRow));
    return {
      id,
      label,
      mods: mods.map(toFeaturedMod)
    };
  });

  const content = { gameId: GAME_ID, categoryRoots, rankings };
  const revision = createHash('sha256').update(JSON.stringify(content)).digest('hex').slice(0, 16);
  // Weekbox's FeaturedService currently accepts schema version 3.
  return { schemaVersion: 3, generatedAt: new Date().toISOString(), revision, ...content };
}

async function readPreviousFeaturedData(fileName) {
  try {
    return JSON.parse(await readFile(new URL(`../public/${fileName}`, import.meta.url), 'utf8'));
  } catch {
    return null;
  }
}

const previousFeaturedData = await readPreviousFeaturedData('featured.json');
const previousVsliceFeaturedData = await readPreviousFeaturedData('featured-vslice.json');
const sources = await fetchFeaturedSources();
const featuredData = await buildFeaturedData(CATEGORY_ROOTS, sources);
if (previousFeaturedData?.revision === featuredData.revision) {
  featuredData.generatedAt = previousFeaturedData.generatedAt;
}
const vsliceFeaturedData = await buildFeaturedData(VSLICE_CATEGORY_ROOTS, sources);
if (previousVsliceFeaturedData?.revision === vsliceFeaturedData.revision) {
  vsliceFeaturedData.generatedAt = previousVsliceFeaturedData.generatedAt;
}
await mkdir(new URL('../public/', import.meta.url), { recursive: true });
await writeFile(new URL('../public/featured.json', import.meta.url), `${JSON.stringify(featuredData, null, 2)}\n`);
await writeFile(new URL('../public/featured-vslice.json', import.meta.url), `${JSON.stringify(vsliceFeaturedData, null, 2)}\n`);
await rm(new URL('../public/featured-manifest.json', import.meta.url), { force: true });

File: scripts/generate-featured.go
package main

import (
	"context"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"
)

const (
	gameID            = 8694
	featuredPerPeriod = 5
	topSubsURL        = "https://gamebanana.com/apiv12/Game/8694/TopSubs"
	profileURL        = "https://gamebanana.com/apiv11/Mod"
	modIndexURL       = "https://gamebanana.com/apiv11/Mod/Index"
)

var categoryRoots = []int{28367, 34764, 3827, 43798, 43850, 43788}
var vsliceCategoryRoots = []int{29202}
var fetchCategoryRoots = uniqueInts(append(append([]int{}, categoryRoots...), vsliceCategoryRoots...))
var excludedModIDs = map[int]bool{309789: true}

var engineByCategory = map[int]EngineInfo{
	29202: {ID: "vslice", Name: "Base Game", Icon: "vslice.png", CategoryName: "Base Game Mod Folders"},
	28367: {ID: "psych", Name: "Psych Engine", Icon: "psych.png", CategoryName: "Psych Engine Mod Folders"},
	34764: {ID: "codename", Name: "Codename Engine", Icon: "codename.png", CategoryName: "Codename Engine Mod Folders"},
	3827:  {ID: "executable", Name: "Executable", Icon: "exe.png", CategoryName: "Executable Mod Folders"},
	43798: {ID: "pslice", Name: "P-Slice", Icon: "pslice.png", CategoryName: "P-Slice Mod Folders"},
	43850: {ID: "fpsplus", Name: "FPS Plus", Icon: "fpsplus.png", CategoryName: "FPS Plus Mod Folders"},
	43788: {ID: "psychonline", Name: "Psych Online", Icon: "psychonline.png", CategoryName: "Psych Online Mod Folders"},
}

var periods = []Period{
	{API: "today", ID: "day", Label: "Best of Today", Seconds: 24 * 60 * 60},
	{API: "week", ID: "week", Label: "Best of This Week", Seconds: 7 * 24 * 60 * 60},
	{API: "month", ID: "month", Label: "Best of This Month", Seconds: 30 * 24 * 60 * 60},
	{API: "3month", ID: "three-months", Label: "Best of 3 Months", Seconds: 90 * 24 * 60 * 60},
	{API: "6month", ID: "six-months", Label: "Best of 6 Months", Seconds: 180 * 24 * 60 * 60},
	{API: "year", ID: "year", Label: "Best of This Year", Seconds: 365 * 24 * 60 * 60},
	{API: "alltime", ID: "all-time", Label: "Best of All Time"},
}

type Period struct {
	API, ID, Label string
	Seconds        int64
}

type EngineInfo struct {
	ID, Name, Icon, CategoryName string
}

type Image struct {
	BaseURL string `json:"_sBaseUrl"`
	File    string `json:"_sFile"`
}

type PreviewMedia struct {
	Images []Image `json:"_aImages"`
}

type Submitter struct {
	Name string `json:"_sName"`
}

type Category struct {
	ID int `json:"_idRow"`
}

type Mod struct {
	ID            int           `json:"_idRow"`
	Name          string        `json:"_sName"`
	ProfileURL    string        `json:"_sProfileUrl"`
	ImageURL      string        `json:"_sImageUrl"`
	Text          string        `json:"_sText"`
	Description   string        `json:"_sDescription"`
	Period        string        `json:"_sPeriod"`
	DateAdded     int64         `json:"_tsDateAdded"`
	DateUpdated   int64         `json:"_tsDateUpdated"`
	DateModified  int64         `json:"_tsDateModified"`
	LikeCount     int64         `json:"_nLikeCount"`
	DownloadCount int64         `json:"_nDownloadCount"`
	ViewCount     int64         `json:"_nViewCount"`
	Submitter     *Submitter    `json:"_aSubmitter"`
	PreviewMedia  *PreviewMedia `json:"_aPreviewMedia"`
	SuperCategory *Category     `json:"_aSuperCategory"`
}

type CategoryMod struct {
	Mod        Mod
	CategoryID int
}

type Sources struct {
	TopSubs       []Mod
	Profiles      map[int]*Mod
	RecentGroups  [][]CategoryMod
	AllTimeGroups [][]CategoryMod
	profileMu     sync.RWMutex
}

type FeaturedMod struct {
	ID          int    `json:"id"`
	Title       string `json:"title"`
	Description string `json:"description"`
	Author      string `json:"author"`
	Image       string `json:"image"`
	Likes       int64  `json:"likes"`
	Downloads   int64  `json:"downloads"`
	Views       int64  `json:"views"`
	PublishedAt int64  `json:"publishedAt"`
	UpdatedAt   int64  `json:"updatedAt"`
	URL         string `json:"url"`
	Engine      struct {
		ID   string `json:"id"`
		Name string `json:"name"`
		Icon string `json:"icon"`
	} `json:"engine"`
	Category struct {
		ID   int    `json:"id"`
		Name string `json:"name"`
	} `json:"category"`
}

type Ranking struct {
	ID    string        `json:"id"`
	Label string        `json:"label"`
	Mods  []FeaturedMod `json:"mods"`
}

type Feed struct {
	SchemaVersion int       `json:"schemaVersion"`
	GeneratedAt   string    `json:"generatedAt"`
	Revision      string    `json:"revision"`
	GameID        int       `json:"gameId"`
	CategoryRoots []int     `json:"categoryRoots"`
	Rankings      []Ranking `json:"rankings"`
}

type feedContent struct {
	GameID        int       `json:"gameId"`
	CategoryRoots []int     `json:"categoryRoots"`
	Rankings      []Ranking `json:"rankings"`
}

type previousFeed struct {
	GeneratedAt string `json:"generatedAt"`
	Revision    string `json:"revision"`
}

var htmlTag = regexp.MustCompile(`<[^>]*>?`)

type generator struct {
	client *http.Client
}

func (g *generator) getJSON(ctx context.Context, endpoint string, target any) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return err
	}
	res, err := g.client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return fmt.Errorf("GameBanana returned %s", res.Status)
	}
	if err := json.NewDecoder(res.Body).Decode(target); err != nil {
		return err
	}
	return nil
}

func (g *generator) fetchTopSubs(ctx context.Context) ([]Mod, error) {
	var raw json.RawMessage
	if err := g.getJSON(ctx, topSubsURL, &raw); err != nil {
		return nil, fmt.Errorf("fetch TopSubs: %w", err)
	}
	trimmed := strings.TrimSpace(string(raw))
	if trimmed == "" || trimmed[0] != '[' {
		return nil, errors.New("GameBanana returned an invalid TopSubs response")
	}
	var mods []Mod
	if err := json.Unmarshal(raw, &mods); err != nil {
		return nil, errors.New("GameBanana returned an invalid TopSubs response")
	}
	return mods, nil
}

func (g *generator) fetchProfile(ctx context.Context, modID int) (*Mod, error) {
	var profile Mod
	if err := g.getJSON(ctx, fmt.Sprintf("%s/%d/ProfilePage", profileURL, modID), &profile); err != nil {
		return nil, fmt.Errorf("fetch profile for mod %d: %w", modID, err)
	}
	return &profile, nil
}

func (g *generator) fetchCategory(ctx context.Context, categoryID int, sortName string, pageLimit int) ([]CategoryMod, error) {
	mods := make([]CategoryMod, 0)
	for page := 1; page <= pageLimit; page++ {
		params := url.Values{}
		params.Set("_sSort", sortName)
		params.Set("_nPage", fmt.Sprint(page))
		params.Set("_nPerpage", "50")
		params.Set("_aFilters[Generic_Game]", fmt.Sprint(gameID))
		params.Set("_aFilters[Generic_Category]", fmt.Sprint(categoryID))

		var response struct {
			Records []Mod `json:"_aRecords"`
		}
		if err := g.getJSON(ctx, modIndexURL+"?"+params.Encode(), &response); err != nil {
			return nil, fmt.Errorf("fetch category %d: %w", categoryID, err)
		}
		for _, mod := range response.Records {
			mods = append(mods, CategoryMod{Mod: mod, CategoryID: categoryID})
		}
		if len(response.Records) < 50 {
			break
		}
	}
	return mods, nil
}

func (g *generator) fetchCategoryGroups(ctx context.Context, categoryIDs []int, sortName string, pageLimit int) ([][]CategoryMod, error) {
	groups := make([][]CategoryMod, len(categoryIDs))
	errs := make(chan error, len(categoryIDs))
	var wg sync.WaitGroup
	for i, categoryID := range categoryIDs {
		wg.Add(1)
		go func(i, categoryID int) {
			defer wg.Done()
			var err error
			groups[i], err = g.fetchCategory(ctx, categoryID, sortName, pageLimit)
			if err != nil {
				errs <- err
			}
		}(i, categoryID)
	}
	wg.Wait()
	close(errs)
	for err := range errs {
		return nil, err
	}
	return groups, nil
}

func (g *generator) fetchSources(ctx context.Context) (*Sources, error) {
	var topSubs []Mod
	var recentGroups, allTimeGroups [][]CategoryMod
	errs := make(chan error, 3)
	var wg sync.WaitGroup
	wg.Add(3)
	go func() {
		defer wg.Done()
		var err error
		topSubs, err = g.fetchTopSubs(ctx)
		if err != nil {
			errs <- err
		}
	}()
	go func() {
		defer wg.Done()
		var err error
		recentGroups, err = g.fetchCategoryGroups(ctx, fetchCategoryRoots, "Generic_NewAndUpdated", 4)
		if err != nil {
			errs <- err
		}
	}()
	go func() {
		defer wg.Done()
		var err error
		allTimeGroups, err = g.fetchCategoryGroups(ctx, fetchCategoryRoots, "Generic_MostLiked", 1)
		if err != nil {
			errs <- err
		}
	}()
	wg.Wait()
	close(errs)
	for err := range errs {
		return nil, err
	}

	profiles := make(map[int]*Mod)
	var profileMu sync.Mutex
	errs = make(chan error, len(uniqueMods(topSubs)))
	var profileWG sync.WaitGroup
	for _, mod := range uniqueMods(topSubs) {
		if excludedModIDs[mod.ID] {
			continue
		}
		mod := mod
		profileWG.Add(1)
		go func() {
			defer profileWG.Done()
			profile, err := g.fetchProfile(ctx, mod.ID)
			if err != nil {
				errs <- err
				return
			}
			profileMu.Lock()
			profiles[mod.ID] = profile
			profileMu.Unlock()
		}()
	}
	profileWG.Wait()
	close(errs)
	for err := range errs {
		return nil, err
	}
	return &Sources{TopSubs: topSubs, Profiles: profiles, RecentGroups: recentGroups, AllTimeGroups: allTimeGroups}, nil
}

func (s *Sources) profile(id int) *Mod {
	s.profileMu.RLock()
	defer s.profileMu.RUnlock()
	return s.Profiles[id]
}

func (s *Sources) setProfile(id int, profile *Mod) {
	s.profileMu.Lock()
	s.Profiles[id] = profile
	s.profileMu.Unlock()
}

func uniqueInts(values []int) []int {
	seen := make(map[int]bool, len(values))
	result := make([]int, 0, len(values))
	for _, value := range values {
		if !seen[value] {
			seen[value] = true
			result = append(result, value)
		}
	}
	return result
}

func uniqueMods(mods []Mod) []Mod {
	indexes := make(map[int]int, len(mods))
	result := make([]Mod, 0, len(mods))
	for _, mod := range mods {
		if index, ok := indexes[mod.ID]; ok {
			result[index] = mod
		} else {
			indexes[mod.ID] = len(result)
			result = append(result, mod)
		}
	}
	return result
}

func uniqueCategoryMods(mods []CategoryMod) []CategoryMod {
	indexes := make(map[int]int, len(mods))
	result := make([]CategoryMod, 0, len(mods))
	for _, entry := range mods {
		if excludedModIDs[entry.Mod.ID] {
			continue
		}
		if index, ok := indexes[entry.Mod.ID]; ok {
			result[index] = entry
		} else {
			indexes[entry.Mod.ID] = len(result)
			result = append(result, entry)
		}
	}
	return result
}

func score(mod Mod) int64 {
	return mod.LikeCount*1_000_000 + mod.DownloadCount*1_000 + mod.ViewCount
}

func activeSince(mod Mod, cutoff int64) bool {
	updated := mod.DateUpdated
	if updated == 0 {
		updated = mod.DateModified
	}
	return mod.DateAdded >= cutoff || updated >= cutoff
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}

func firstNonZero(values ...int64) int64 {
	for _, value := range values {
		if value != 0 {
			return value
		}
	}
	return 0
}

func imageURL(mod Mod) string {
	if mod.ImageURL != "" {
		return mod.ImageURL
	}
	if mod.PreviewMedia != nil && len(mod.PreviewMedia.Images) > 0 {
		image := mod.PreviewMedia.Images[0]
		return image.BaseURL + "/" + image.File
	}
	return "https://images.gamebanana.com/img/ss/mods/default.jpg"
}

func cleanDescription(mod Mod, profile *Mod) string {
	profileText, profileDescription := "", ""
	if profile != nil {
		profileText, profileDescription = profile.Text, profile.Description
	}
	text := firstNonEmpty(profileText, mod.Text, mod.Description, profileDescription)
	return strings.Join(strings.Fields(htmlTag.ReplaceAllString(text, "")), " ")
}

func toFeaturedMod(entry CategoryMod, profile *Mod) FeaturedMod {
	mod := entry.Mod
	engine := engineByCategory[entry.CategoryID]
	result := FeaturedMod{
		ID:          mod.ID,
		Title:       mod.Name,
		Description: cleanDescription(mod, profile),
		Author:      "Unknown",
		Image:       imageURL(mod),
		Likes:       firstNonZero(mod.LikeCount, profileValue(profile, func(p *Mod) int64 { return p.LikeCount })),
		Downloads:   profileValue(profile, func(p *Mod) int64 { return p.DownloadCount }),
		Views:       profileValue(profile, func(p *Mod) int64 { return p.ViewCount }),
		PublishedAt: profileValue(profile, func(p *Mod) int64 { return p.DateAdded }),
		UpdatedAt:   firstNonZero(profileValue(profile, func(p *Mod) int64 { return p.DateUpdated }), profileValue(profile, func(p *Mod) int64 { return p.DateModified })),
		URL:         firstNonEmpty(mod.ProfileURL, profileValueString(profile, func(p *Mod) string { return p.ProfileURL }), fmt.Sprintf("https://gamebanana.com/mods/%d", mod.ID)),
	}
	if mod.Submitter != nil && mod.Submitter.Name != "" {
		result.Author = mod.Submitter.Name
	}
	result.Engine.ID, result.Engine.Name, result.Engine.Icon = engine.ID, engine.Name, engine.Icon
	result.Category.ID, result.Category.Name = entry.CategoryID, engine.CategoryName
	return result
}

func profileValue(profile *Mod, value func(*Mod) int64) int64 {
	if profile == nil {
		return 0
	}
	return value(profile)
}

func profileValueString(profile *Mod, value func(*Mod) string) string {
	if profile == nil {
		return ""
	}
	return value(profile)
}

func loadFallbackProfiles(ctx context.Context, g *generator, sources *Sources, entries []CategoryMod) []CategoryModProfile {
	result := make([]CategoryModProfile, len(entries))
	var wg sync.WaitGroup
	for i, entry := range entries {
		i, entry := i, entry
		wg.Add(1)
		go func() {
			defer wg.Done()
			profile := sources.profile(entry.Mod.ID)
			if profile == nil {
				var err error
				profile, err = g.fetchProfile(ctx, entry.Mod.ID)
				if err != nil {
					profile = &entry.Mod
				} else {
					sources.setProfile(entry.Mod.ID, profile)
				}
			}
			result[i] = CategoryModProfile{Entry: entry, Profile: profile}
		}()
	}
	wg.Wait()
	return result
}

type CategoryModProfile struct {
	Entry   CategoryMod
	Profile *Mod
}

func buildFeaturedData(ctx context.Context, g *generator, categoryRoots []int, sources *Sources) (Feed, error) {
	allowed := make(map[int]bool, len(categoryRoots))
	for _, categoryID := range categoryRoots {
		allowed[categoryID] = true
	}

	eligible := make([]CategoryMod, 0)
	for _, mod := range uniqueMods(sources.TopSubs) {
		if excludedModIDs[mod.ID] {
			continue
		}
		profile := sources.profile(mod.ID)
		if profile != nil && profile.SuperCategory != nil && allowed[profile.SuperCategory.ID] {
			eligible = append(eligible, CategoryMod{Mod: mod, CategoryID: profile.SuperCategory.ID})
		}
	}
	var recent, allTime []CategoryMod
	for _, group := range sources.RecentGroups {
		for _, entry := range group {
			if allowed[entry.CategoryID] {
				recent = append(recent, entry)
			}
		}
	}
	for _, group := range sources.AllTimeGroups {
		for _, entry := range group {
			if allowed[entry.CategoryID] {
				allTime = append(allTime, entry)
			}
		}
	}
	recent, allTime = uniqueCategoryMods(recent), uniqueCategoryMods(allTime)

	now := time.Now().Unix()
	featuredIDs := make(map[int]bool)
	rankings := make([]Ranking, 0, len(periods))
	for _, period := range periods {
		primary := make([]CategoryMod, 0)
		for _, entry := range eligible {
			if entry.Mod.Period == period.API && !featuredIDs[entry.Mod.ID] {
				primary = append(primary, entry)
			}
		}
		selected := make(map[int]bool, len(featuredIDs)+len(primary))
		for id := range featuredIDs {
			selected[id] = true
		}
		for _, entry := range primary {
			selected[entry.Mod.ID] = true
		}

		candidates := recent
		if period.Seconds == 0 {
			candidates = allTime
		}
		cutoff := now - period.Seconds
		fallback := make([]CategoryMod, 0)
		for _, entry := range candidates {
			if !selected[entry.Mod.ID] && (period.Seconds == 0 || activeSince(entry.Mod, cutoff)) {
				fallback = append(fallback, entry)
			}
		}
		sort.SliceStable(fallback, func(i, j int) bool {
			left, right := fallback[i].Mod, fallback[j].Mod
			if score(left) != score(right) {
				return score(left) > score(right)
			}
			return left.ID > right.ID
		})
		needed := featuredPerPeriod - len(primary)
		if needed < 0 {
			needed = 0
		}
		if len(fallback) > needed {
			fallback = fallback[:needed]
		}
		fallbackProfiles := loadFallbackProfiles(ctx, g, sources, fallback)

		mods := make([]FeaturedMod, 0, len(primary)+len(fallbackProfiles))
		for _, entry := range primary {
			mods = append(mods, toFeaturedMod(entry, sources.profile(entry.Mod.ID)))
			featuredIDs[entry.Mod.ID] = true
		}
		for _, entry := range fallbackProfiles {
			mods = append(mods, toFeaturedMod(entry.Entry, entry.Profile))
			featuredIDs[entry.Entry.Mod.ID] = true
		}
		rankings = append(rankings, Ranking{ID: period.ID, Label: period.Label, Mods: mods})
	}

	content := feedContent{GameID: gameID, CategoryRoots: categoryRoots, Rankings: rankings}
	encoded, err := json.Marshal(content)
	if err != nil {
		return Feed{}, err
	}
	hash := sha256.Sum256(encoded)
	feed := Feed{
		SchemaVersion: 3,
		GeneratedAt:   time.Now().UTC().Format("2006-01-02T15:04:05.000Z"),
		Revision:      fmt.Sprintf("%x", hash[:])[:16],
		GameID:        content.GameID,
		CategoryRoots: content.CategoryRoots,
		Rankings:      content.Rankings,
	}
	if err := validateFeed(feed, allowed); err != nil {
		return Feed{}, err
	}
	return feed, nil
}

func validateFeed(feed Feed, allowed map[int]bool) error {
	for _, ranking := range feed.Rankings {
		if len(ranking.Mods) > featuredPerPeriod {
			return fmt.Errorf("ranking %s has too many mods", ranking.ID)
		}
		for _, mod := range ranking.Mods {
			if !allowed[mod.Category.ID] {
				return fmt.Errorf("mod %d has category %d outside feed", mod.ID, mod.Category.ID)
			}
		}
	}
	return nil
}

func readPrevious(path string) previousFeed {
	data, err := os.ReadFile(path)
	if err != nil {
		return previousFeed{}
	}
	var previous previousFeed
	if json.Unmarshal(data, &previous) != nil {
		return previousFeed{}
	}
	return previous
}

func writeFeed(path string, feed Feed) error {
	data, err := json.MarshalIndent(feed, "", "  ")
	if err != nil {
		return err
	}
	data = append(data, '\n')
	return os.WriteFile(path, data, 0644)
}

func main() {
	ctx := context.Background()
	g := &generator{client: &http.Client{Timeout: 60 * time.Second}}
	sources, err := g.fetchSources(ctx)
	if err != nil {
		panic(err)
	}

	featured, err := buildFeaturedData(ctx, g, categoryRoots, sources)
	if err != nil {
		panic(err)
	}
	if previous := readPrevious("public/featured.json"); previous.Revision == featured.Revision {
		featured.GeneratedAt = previous.GeneratedAt
	}

	vslice, err := buildFeaturedData(ctx, g, vsliceCategoryRoots, sources)
	if err != nil {
		panic(err)
	}
	if previous := readPrevious("public/featured-vslice.json"); previous.Revision == vslice.Revision {
		vslice.GeneratedAt = previous.GeneratedAt
	}

	if err := os.MkdirAll("public", 0755); err != nil {
		panic(err)
	}
	if err := writeFeed("public/featured.json", featured); err != nil {
		panic(err)
	}
	if err := writeFeed("public/featured-vslice.json", vslice); err != nil {
		panic(err)
	}
	if err := os.Remove("public/featured-manifest.json"); err != nil && !errors.Is(err, os.ErrNotExist) {
		panic(err)
	}
}
